### PR TITLE
Megatron Encoder-Decoder Sampler Function

### PR DESCRIPTION
**Pull request by michalivne**
# What does this PR do ?

This PR adds support in external sampler function to Megatron encoder-decoder.

**Collection**: [Note which collection this PR will affect]

# Changelog 

1. Added support in a custom token sampling function to `MegatronLMEncoderDecoderModel.decode`.
2. Added greedy and top k / top p token sampling methods.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by MaximumEntropy — COMMENTED** [review]
LGTM!

**Review by MaximumEntropy — APPROVED** [review]

**Review by MaximumEntropy — APPROVED** [review]

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -1223,14 +1244,17 @@ def dummy():
                         dim=-1, index=torch.tensor(ignore_ids, device=device), value=-float('Inf')
                     )
 
-                # TODO: do log_softmax in fp32?
-                log_probs, token_ids = torch.max(torch.nn.functional.log_softmax(output_tensor, dim=-1), dim=-1)
+                log_probs, token_ids = sample_token_fn(logits=output_tensor[:, -1, :])
+                # enforce valid range of token ids
+                token_ids = torch.clamp(token_ids, max=tokenizer.vocab_size - 1)
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1249]: Not related to this PR, but this seems like a weird use of clamp. Should a token_ids > vocab raise an error? @ericharper

Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:1249]: It can happen due to vocabulary padding (mostly in the early stages of training).
